### ISSUE TITLE
bbin install should not overwrite other executables in bin - potential security issue

### ISSUE THREAD

**Issue by ieugen** [babashka/bbin#40]
bbin install should fail when a binary with the same name exists at the path. 
It should IMO allow user to force installation  with a flag.

This could be a security issue, especially with package owners allowed to set the binary name. 
Some people might chose to override things like `ls` or other stuff.  

Current status:
* bbin installs binary commands as they are
* If a binary is present it will be overwritten

```bash
~/.local/bin$ ls -lah neil
-rwxr-xr-x 1 ieugen ieugen 1,2K sep 27 11:05 neil

BABASHKA_BBIN_DIR=~/.local bbin install io.github.babashka/bbin --latest-sha --as neil
{:coords
 {:git/url "https://github.com/babashka/bbin",
  :git/sha "6fde1b1dbfaef3063eb1eba4899a730bf703c792"},
 :lib io.github.babashka/bbin}

~/.local/bin$ ls -lah neil
-rwxr-xr-x 1 ieugen ieugen 1,1K sep 27 11:26 neil

cat neil 
#!/usr/bin/env bb

; :bbin/start
;
; {:coords
;  {:git/url "https://github.com/babashka/bbin",
;   :git/sha "6fde1b1dbfaef3063eb1eba4899a730bf703c792"},
;  :lib io.github.babashka/bbin}
;
; :bbin/end

(require '[babashka.process :as process]
         '[babashka.fs :as fs]
         '[clojure.string :as str])

(def script-root "/home/ieugen/.gitlibs/libs/io.github.babashka/bbin/6fde1b1dbfaef3063eb1eba4899a730bf703c792")
(def script-lib 'io.github.babashka/bbin)
```

Might be related to 
* https://github.com/babashka/bbin/issues/34 
* https://github.com/babashka/bbin/issues/18
* https://github.com/babashka/bbin/issues/35

**Comment by borkdude** [comment]
I think it's a feature that the older version gets overwritten with the newer version.

**Comment by borkdude** [comment]
We could however prevent overwriting by introducing a flag: `--force` or so

**Comment by rads** [comment]
@ieugen: If you're concerned about security, you need to make sure you only install scripts from a) authors you trust or b) code that you've verified yourself. If you're installing from an untrusted source, there's nothing that `bbin` can do to help you since we don't have any sandboxing feature in Babashka. Blocking script names isn't enough on its own since the script can do whatever it wants with the filesystem when it's running.

**Comment by rads** [comment]
I also recommend putting local bin directories as a _suffix_ in the `PATH` rather than a _prefix_. This is how the README does it:
```
echo 'export PATH="$PATH:$HOME/.babashka/bbin/bin"' >> ~/.$(basename $SHELL)rc && exec $SHELL
```
This reduces the risk of overriding something like `sudo` if you did happen to install a malicious script.

**Comment by ieugen** [comment]
@borkdude : 
> I think it's a feature that the older version gets overwritten with the newer version. 

I can overwrite ANY binary this way so IMO it needs some consideration or at least a notice.

@rads :
> export PATH="$PATH:$HOME/.local/bin" 

This is normally set by the distribution.

> Blocking script names during bbin install isn't enough on its own since the script can do whatever it wants with the filesystem when it's running.

True. We can maybe do one or more of the following:
* print the changes that are being done what apps are installed and where
* ask for permission, esp in case where it will overwrite an existing binary. This could be done on first install only if we consider the next ones to be upgrades and overwriting is expected.
* If the script is linked in /local/bin we can check it's contents hash match what we expect to have, before we attempt changes ?!

**Comment by jeroenvandijk** [comment]
I agree with @ieugen. I think leaving **all** the responsibility to the end user of `bbin` is simply not fair. This user has to understand the following:

- What a Babashka script can do (file writing, http requests etc)
- How `bbin` will use the coordinates to create a binary
- What potential other binaries this new binary is overwriting, literally or by PATH precedence.
- What this other binary is responsible for.

I believe something like `bbin` should exist for ease of distribution so moving the responsibility of all these safety concerns removes the ease for a serious user actually. Ease would only exist for the naive user, until things go wrong and no one will trust `bbin` again. Sounds a bit dramatic, but this is how trust works I believe.

So one extra vote for having `bbin` do a maximum effort to protect the end-user, as this IMO will really contribute to the distribution.

I would want this to go even further than only `protecting existing binaries`, I would also like so see:

- Telling the user what the script potentially can do. Does it write to non-temporary dirs (analyzing the filesystem code would do a great deal here)
- Does the script do external requests
- etc.

Maybe a permission system would be interesting. "Script asks permission for X, Y, Z. Please confirm"
